sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/StateBackedIterable.java

```diff
@@ -0,0 +1,223 @@
+ * CoderTranslator#fromComponents} to be able to create a {@link StateBackedIterable.Coder}.
+ */
+public class StateBackedIterable<T> implements Iterable<T> {
+
+  private final BeamFnStateClient beamFnStateClient;
+  private final org.apache.beam.sdk.coders.Coder<T> elemCoder;
+  @VisibleForTesting final StateRequest request;
+  @VisibleForTesting final List<T> prefix;
+
+  public StateBackedIterable(
+      BeamFnStateClient beamFnStateClient,
+      String instructionId,
+      ByteString runnerKey,
+      org.apache.beam.sdk.coders.Coder<T> elemCoder,
+      List<T> prefix) {
+    this.beamFnStateClient = beamFnStateClient;
+    this.elemCoder = elemCoder;
+
+    StateRequest.Builder requestBuilder = StateRequest.newBuilder();
+    requestBuilder
+        .setInstructionId(instructionId)
+        .getStateKeyBuilder()
+        .getRunnerBuilder()
+        .setKey(runnerKey);
+    this.request = requestBuilder.build();
+    this.prefix = prefix;
+  }
+
+  @Override
+  public Iterator<T> iterator() {
+    return Iterators.concat(
+        prefix.iterator(),
+        new DataStreams.DataStreamDecoder(
+            elemCoder,
+            DataStreams.inbound(StateFetchingIterators.forFirstChunk(beamFnStateClient, request))));
+  }
+
+  /**
+   * Decodes an {@link Iterable} that might be backed by state. If the terminator at the end of the
+   * value stream is {@code -1} then we return a {@link StateBackedIterable} otherwise we return an
+   * {@link Iterable}.
+   */
+  public static class Coder<T> extends IterableLikeCoder<T, Iterable<T>> {
+
+    private final BeamFnStateClient beamFnStateClient;
+    private final Supplier<String> instructionId;
+
+    public Coder(
+        BeamFnStateClient beamFnStateClient,
+        Supplier<String> instructionId,
+        org.apache.beam.sdk.coders.Coder<T> elemCoder) {
+      super(elemCoder, "StateBackedIterable");
+      this.beamFnStateClient = beamFnStateClient;
+      this.instructionId = instructionId;
+    }
+
+    @Override
+    protected Iterable<T> decodeToIterable(List<T> decodedElements) {
+      return decodedElements;
+    }
+
+    @Override
+    protected Iterable<T> decodeToIterable(
+        List<T> decodedElements, long terminatorValue, InputStream in) throws IOException {
+      if (terminatorValue == -1L) {
+        long tokenLength = VarInt.decodeLong(in);
+        ByteString token = ByteString.readFrom(ByteStreams.limit(in, tokenLength));
+        return new StateBackedIterable<>(
+            beamFnStateClient, instructionId.get(), token, getElemCoder(), decodedElements);
+      } else {
+        throw new IllegalStateException(
+            String.format(
+                "StateBackedIterable expected terminator of 0 or -1 but received %s.",
+                terminatorValue));
+      }
+    }
+
+    @Override
+    public void encode(Iterable<T> iterable, OutputStream outStream) throws IOException {
```

Comment:
I am guessing the encoding for state channel also exist already in Java SDK.  That is why we don't need to implement them in this PR. Would you please point me to that?

Reply:
See https://github.com/apache/beam/blob/master/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/BeamFnStateClient.java and https://github.com/apache/beam/blob/master/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/BeamFnStateGrpcClientCache.java

Reply:
Thanks!